src/ModelContextProtocol/Protocol/Transport/StdioServerTransport.cs

```diff
@@ -137,6 +138,8 @@ public Task StartListeningAsync(CancellationToken cancellationToken = default)
     /// <inheritdoc/>
     public override async Task SendMessageAsync(IJsonRpcMessage message, CancellationToken cancellationToken = default)
     {
+        using var _ = await _sendLock.LockAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
The crux of https://github.com/modelcontextprotocol/csharp-sdk/issues/1 is that this method is racing with the background message reader which disconnects the transport on line 290. I think it should be possible to address that issue if we guard that line with the same semaphore.